source/blog/_posts/2017-11-17-on-testing-strategies.md

```diff
@@ -0,0 +1,282 @@
+
+## The Application
+
+First we take a look at a fairly standard application design. We need this to have a shared base for the comparision of the different testing strategies. This model omits the framework and data structures which are for our not essential for our evaluation.
+
+We have a controller, that uses a domain service and two auxiliary services to handle requests. The services themselves have dependencies on other services or storage implementations.
+
+<img src="/blog/img/advantages-of-integration-testing/standard-application.svg" />
+
+So how should we test this? Lt's discuss some strategies:
+
+## Testing Strategy: Functional testing only
+
+Many projects out there use acceptance tests as their only way of automated testing. So why is it a bad idea? One word: Performance! Functional tests use a real browser to check the application. Even high performance applications have response times of at least 50ms, excluding the JavaScript execution time you are quickly at 500ms or more - for a **single page**. So let's math with an average of 250ms per page:
+
+| Page Loads  |  Time   |
+|-------------|---------|
+| 1           | 250 ms  |
+| 5           |   1 min |
+| 10          |   2 min |
+| 50          |  10 min |
+| 100         |  20 min |
+| 400         |   1 hr  |
+
+Now you have to ask yourself: Do you want to maintain this? Is it useful? Creating an application that has 400 different click paths is quite easy, but waiting an hour to validate every single change is tedious at least and at worst it stops being helpful quickly.
+
+The main advantage is that - at least for settled applications - introducing testing late in a development flow makes this a quick win because usually you have to write the least amount of code for this. And even if the Structure of your source code does not allow to easily introduce testing into every layer, these tests at least tend to solidify the basic functionality. So this is possible by just introducing a single layer on top of your application:
+
+<img src="/blog/img/advantages-of-integration-testing/application-functional-testing.svg" />
+
+**Good:**
+
+* Gives you 99% certainty that the application works correctly.
+* Adds minimal code to maintain.
+
+**Bad:**
+
+* Takes a long time even for small applications
+* Does not help developers
+* Does nothing to your application architecture, but add a few css classes
+* Usually very fragile, small changes tend to have an unreasonably big impact (you change a buttons position the login breaks)
+
+**Verdict:** Run!
+
+## Testing Strategy: The Pyramid
+
+There is a testing pyramid favored by people like Robert C. Martin, it looks like this:
+
+<img src="/blog/img/advantages-of-integration-testing/unit-test-pyramid.svg" />
+
+The Testing Pyramid puts the focus on **unit tests** proposes to write a few integration tests and if really necessary very few functional test. So for our example application this looks like this:
```

Comment:
Maybe better : "A Testing Pyramid _that_ puts the focus on unit tests . . .  few functional tests"